can/bit_timing.py

```diff
@@ -0,0 +1,213 @@
+            point to the end of the bit.
+        :param int sjw:
+            The Synchronization Jump Width. Decides the maximum number of time quanta
+            that the controller can resynchronize every bit.
+        :param int nof_samples:
+            Either 1 or 3. Some CAN controllers can also sample each bit three times.
+            In this case, the bit will be sampled three quanta in a row,
+            with the last sample being taken in the edge between TSEG1 and TSEG2.
+            Three samples should only be used for relatively slow baudrates.
+        :param int btr0:
+            The BTR0 register value used by many CAN controllers.
+        :param int btr1:
+            The BTR1 register value used by many CAN controllers.
+        """
+        self._bitrate = bitrate
+        self._brp = brp
+        self._sjw = sjw
+        self._tseg1 = tseg1
+        self._tseg2 = tseg2
+        self._nof_samples = nof_samples
+        self._f_clock = f_clock
+
+        if btr0 is not None:
+            self._brp = (btr0 & 0x3F) + 1
+            self._sjw = (btr0 >> 6) + 1
+        if btr1 is not None:
+            self._tseg1 = (btr1 & 0xF) + 1
+            self._tseg2 = ((btr1 >> 4) & 0x7) + 1
+            self._nof_samples = 3 if btr1 & 0x80 else 1
+
+        if nof_samples not in (1, 3):
+            raise ValueError("nof_samples must be 1 or 3")
+
+    @property
+    def nbt(self) -> int:
+        """Nominal Bit Time."""
+        return self.sync_seg + self.tseg1 + self.tseg2
+
+    @property
+    def bitrate(self) -> Union[int, float]:
+        """Bitrate in bits/s."""
+        if self._bitrate:
+            return self._bitrate
+        if self._f_clock and self._brp:
+            return self._f_clock / (self._brp * self.nbt)
+        raise ValueError("bitrate must be specified")
+
+    @property
+    def brp(self) -> int:
+        """Bit Rate Prescaler."""
+        if self._brp:
+            return self._brp
+        if self._f_clock and self._bitrate:
+            return round(self._f_clock / (self._bitrate * self.nbt))
+        raise ValueError("Either bitrate and f_clock or brp must be specified")
+
+    @property
+    def sjw(self) -> int:
+        """Synchronization Jump Width."""
+        if not self._sjw:
+            raise ValueError("sjw must be specified")
+        return self._sjw
+
+    @property
+    def tseg1(self) -> int:
+        """Time segment 1.
+
+        The number of quanta from (but not including) the Sync Segment to the sampling point.
+        """
+        if not self._tseg1:
+            raise ValueError("tseg1 must be specified")
+        return self._tseg1
+
+    @property
+    def tseg2(self) -> int:
+        """Time segment 2.
+
+        The number of quanta from the sampling point to the end of the bit.
+        """
+        if not self._tseg2:
+            raise ValueError("tseg2 must be specified")
+        return self._tseg2
+
+    @property
+    def nof_samples(self) -> int:
```

Comment:
Consider just `samples`?

Reply:
Maybe it’s just me but that makes me think it is a list of samples rather than an integer.

I was also thinking about using a bool for using three samples or not.